source/extensions/priv/server/ntds_jet.c

```diff
@@ -0,0 +1,363 @@
+/*!
+* @file ntds_jet.c
+* @brief Definitions for NTDS Jet Engine functions
+*/
+#include "precomp.h"
+
+/*!
+* @brief Shuts down the Jet Instance and frees the jetState struct.
+* @param ntdsState Pointer to a jetsState struct which contains all the state data for the Jet Instance.
+* @returns Indication of sucess or failure.
+*/
+JET_ERR engine_shutdown(jetState *ntdsState){
+	JET_ERR shutdownStatus;
+	shutdownStatus = JetCloseDatabase(ntdsState->jetSession, ntdsState->jetDatabase, (JET_GRBIT)NULL);
+	if (shutdownStatus != JET_errSuccess){
+		return shutdownStatus;
+	}
+	shutdownStatus = JetDetachDatabase(ntdsState->jetSession, ntdsState->ntdsPath);
+	if (shutdownStatus != JET_errSuccess){
+		return shutdownStatus;
+	}
+	shutdownStatus = JetEndSession(ntdsState->jetSession, (JET_GRBIT)NULL);
+	if (shutdownStatus != JET_errSuccess){
+		return shutdownStatus;
+	}
+	shutdownStatus = JetTerm(ntdsState->jetEngine);
+	free(ntdsState);
+	return shutdownStatus;
+}
+
+/*!
+* @brief Starts up the Jet Instance and initialises it.
+* @param ntdsState Pointer to a jetsState struct which contains all the state data for the Jet Instance.
+* @returns Indication of sucess or failure.
+*/
+JET_ERR engine_startup(jetState *ntdsState){
+	JET_ERR jetError;
+	// Set the Page Size to the highest possibile limit
+	jetError = JetSetSystemParameter(&ntdsState->jetEngine, JET_sesidNil, JET_paramDatabasePageSize, 8192, NULL);
+	if (jetError != JET_errSuccess){
+		return jetError;
+	}
+	// Create our Jet Instance
+	jetError = JetCreateInstance(&ntdsState->jetEngine, "NTDS");
+	if (jetError != JET_errSuccess){
+		return jetError;
+	}
+	// Disable crash recovery and transaction logs
+	jetError = JetSetSystemParameter(&ntdsState->jetEngine, JET_sesidNil, JET_paramRecovery, (JET_API_PTR)NULL, "Off");
+	if (jetError != JET_errSuccess){
+		return jetError;
+	}
+	// Initialise the Jet instance
+	jetError = JetInit(&ntdsState->jetEngine);
+	if (jetError != JET_errSuccess){
+		return jetError;
+	}
+	return JET_errSuccess;
+}
+
+/*!
+* @brief Moves the database cursor to the first record in the 'datatable' table
+* @param ntdsState Pointer to a jetsState struct which contains all the state data for the Jet Instance.
+* @returns Indication of sucess or failure.
+*/
+JET_ERR find_first(jetState *ntdsState){
+	JET_ERR cursorStatus;
+	cursorStatus = JetMove(ntdsState->jetSession, ntdsState->jetTable, JET_MoveFirst, (JET_GRBIT)NULL);
+	return cursorStatus;
+}
+
+/*!
+* @brief Collect the Column Definitions for all relevant columns in 'datatable'
+* @param ntdsState Pointer to a jetsState struct which contains all the state data for the Jet Instance.
+* @param accountColumns Pointer to an ntdsState struct which will hold all of our column definitions.
+* @returns Indication of sucess or failure.
+*/
+JET_ERR get_column_info(jetState *ntdsState, ntdsColumns *accountColumns){
+	JET_ERR columnError;
+	const char attributeNames[][25] = {

```

Comment:
you can make this shorter and easier to read by declaring an anonymous struct instead:
```
   struct {
       char *name;
       JETCOLUMNDEF *column;
   } columns[] = {
       {"ATTm590045", &accountColumns->accountName},
       {"ATTj590126", &accountColumns->accountType}
...
   };
```
Then you can iterate over the struct. You also don't need to declare a string length when referring to a literal. Char * will do just fine.